src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs

```diff
@@ -112,8 +116,17 @@ public virtual string NewLine
             {
                 value ??= Environment.NewLineConst;
 
-                CoreNewLineStr = value;
-                CoreNewLine = value.ToCharArray();
+                if (CoreNewLineStr != value)
+                {
+                    CoreNewLineStr = value;
+                    CoreNewLine =
+                        value == Environment.NewLineConst ? s_coreNewLine : // current OS default
+                        Environment.NewLineConst == "\r\n" && value == "\n" ? (s_otherCoreNewLine ??= ['\n']) : // other OS default
+                        Environment.NewLineConst == "\n" && value == "\r\n" ? (s_otherCoreNewLine ??= ['\r', '\n']) : // other OS default
+                        value.ToCharArray(); // unknown
```

Comment:
Maybe a switch expression to improve readability?

```suggestion
                    CoreNewLine = (value, Environment.NewLineConst) switch
                    {
                        // current OS default
                        (Environment.NewLineConst, Environment.NewLineConst) => s_coreNewLine,

                        // other OS default
                        ("\n", "\r\n") => s_otherCoreNewLine ??= ['\n'],
                        ("\r\n", "\n") => s_otherCoreNewLine ??= ['\r', '\n'],

                        // unknown
                        _ => value.ToCharArray(),
                    };
```

Reply:
Ah, I see. With tuples Roslyn doesn't see that the branche is dead code.

With the following it does [Sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQ64GYAEicDCOA3ujuTgJYB2wApgE7VgA2OAxgPbUTB5IAGHADk6AdwAyNOgW68cAXhwAiADoNV1ZQG4yFPeQAODTvXb0AJhwAWYBgG0AuoU4M6oydMU4IAfS5uHlLUdLpoFDjGlABuYPT8AGw4bmAW3CwAnjZ2Tj7+ru7iwaEGkQwxcXSJ2Q6OAPx5ptaMsoFF0mERpdh4ACw4AIIAFHCCOLEsAK50AJSlpOER5K2FniGKpUsUE9OKSkHSsjx8DX4Bq8U4IDgA9DcckwxutDgA8gDKOBZ0AGZgkyxgJstiJ2iEjvIFEo1BotDgAGTw8asXZQlSaZQ4BpDPxNFoFA7rOp1JT2ADkmjJjhmV1u9zxDDen2+fwBQMWIIohJkcj4aLUcMRyKmVX56gxWJwON8DJW3KxJJw5PUZIANDgKdQqTTrnccAymV9fv9AcCtjs6AA6AAqnAItgYAyeYAyQxm2jpOEm1AA1tROGJqKUAL6lbq4OD9ABCIzGFrmHJIZpcbTWouFuwgYkowHY1mTC05ET1Bo+RtZpsTnIFmLEzWooLTEL50PFcIUAD5GsBmgw5WCqsTSZqqarkxEYRK63QG9zm3t0e2u7ie/jU5ch0qKQw1RrKY4x1WQXr2I9nnwyyyTeyi1yB/PO3lztzD7fyHrvX6A0Gj1tfIouwtG07QdJ0GBdN1XxBYNOgoUM0GDIA===) / [DiffChecker](https://www.diffchecker.com/6P9AHjAX/).

```suggestion
                    CoreNewLine = value switch
                    {
                        // other OS default
                        "\n" when Environment.NewLineConst == "\r\n" => s_otherCoreNewLine ??= ['\n'],
                        "\r\n" when Environment.NewLineConst == "\n" => s_otherCoreNewLine ??= ['\r', '\n'],
                        // current OS default
                        Environment.NewLineConst => s_coreNewLine,
                        // unknown
                        _ => value.ToCharArray(),
                    };
```

Reply:
> That doesn't currently happen with the switch variant.

File an issue?